### ISSUE TITLE
Content changes disappear in visual editor after saving 

### ISSUE THREAD

**Issue by yvonnetangsu** [storyblok/gatsby-source-storyblok#194]
[storyblokurl]: https://www.storyblok.com?utm_source=github.com&utm_medium=issue-template

---

## Expected Behavior

After setting up the site following the Gatsby 5 minute tutorial on the Storyblok website, I should be able to see changes I made in the Storyblok v2 visual editor even after saving.

## Current Behavior

While I'm able to see my changes as I'm typing into Storyblok, as soon as I click the "Save" button, my changes disappears in the visual editor and revert to the initial state when I started `npm run develop`.

## Steps to Reproduce

1. Follow the Gatsby 5 minute tutorial on the Storyblok website
2. Make changes on a page on Storyblok, eg, the homepage, then click "Save" and see that your changes go away in the visual editor.

**Comment by schabibi1** [comment]
Hi @yvonnetangsu 

Could you tell me if the content changes would be loaded when you set the "Refresh Data" button on the GraohiQL playground?
Follow this part of the tutorial: [https://www.storyblok.com/tp/add-a-headless-cms-to-gatsby-5-minutes#using-the-graphiql-explorer](https://www.storyblok.com/tp/add-a-headless-cms-to-gatsby-5-minutes#using-the-graphiql-explorer)

When you set this npm script and hit the "Refresh data" button on the GraphiQL playground, it should load when you serve pages with SSG.

```
  "scripts": {
     ...
    "develop": "ENABLE_GATSBY_REFRESH_ENDPOINT=true gatsby develop",
     ...
}
```

Also, please keep in mind that if you serve content with SSG, the most updated content will be loaded at build time. If you'd like to load the most updated content changes whenever you hit the Save button on the Storyblok visual editor, serving content with SSR will allow you to do so.

Gatsby docs, SSR: [https://www.gatsbyjs.com/docs/how-to/rendering-options/using-server-side-rendering/](https://www.gatsbyjs.com/docs/how-to/rendering-options/using-server-side-rendering/)

**Comment by yvonnetangsu** [comment]
Hi @schabibi1 !

I tried clicking the "refresh data" button in the GraphiQL playground, and it does help a bit but I don't think it is working as intended still. I'll describe what I did step by step below

**Case 1 - not clicking "Refresh data" in the GraphiQL playground.**

1. I have added the "ENABLE_GATSBY_REFRESH_ENDPOINT=true" flag to my "develop" script. 
2. I do `npm run develop` and started the https proxy. 
3. As I click into the, eg, home page on Storyblok, I see that I'm connecting to the visual editor with the content that was previously saved (when I started `npm run develop`). Let's say, the headline at this time is "Hello"
4. I start typing in Storyblok headline field and add "oooo" to the end of "Hello". I'm seeing the visual editor updating in real time and it's showing "Hellooooo" both in the headline field, and in the editor. This is great.
5. I click the "Save" button on Storyblok, and immediately, in the visual editor, my headline went back to being just "Hello" while the headline field still keeps my change and is displayed correctly as "Hellooooo"
6. I click the "reload" icon button in Storyblok (the curved arrow button), and I'm still seeing "Hello" in the visual editor.
7. The only way for me to see my saved changes (Hellooooo in headline) is to manually restart `npm run develop`

**Case 2 - clicking "Refresh data" in the GraphiQL playground.**

Step 1-5 is the same as in Case 1.
6. I click the "Refresh data" button in the GraphiQL playground.
7. I go back to Storyblok and click the "reload" icon button (the curved arrow button).
8. This time, I do see the changes that I saved, ie "Hellooooo" in the visual editor.

So clicking the "Refresh data" in the GraphiQL playground definitely helps, but this doesn't seem intuitive for content editors for example. I recall when using the old boilerplate (back when using Storyblok editor v1), after I click "Save", my changes don't disappear in the visual editor. Let's say I'm using Microsoft Word, when I start typing and save, I would expect the document to stay the same in the state that I save it in, and not reverting back to the state when I open the document 🤔 

I've been going through the Discord Gatsby channel, and I see a few others having the same issue, eg
https://discord.com/channels/700316478792138842/700326489731891201/1007650315325145153
https://discord.com/channels/700316478792138842/700326489731891201/1027638855630856222

Some mentioned that the issue is with the `gatsby-source-storyblok` plugin and they have to write their own code and not use the plugin because of this issue.

Would appreciate any insight on this - thanks in advance!

**Comment by petr166** [comment]
I have the same issue and from looking around it seems to be a 2 part problem:

1. When pressing "Save" or "Publish" a window reload is triggered as defined in the [original event handler in `storyblok-js`](https://github.com/storyblok/storyblok-js/blob/8c887c3c35950bb8070abb13417d4f8742437b69/lib/index.ts#L48)
    - removing the case for "change" & "published" leaves the preview untouched upon saving or publishing in Storyblok UI
    - this doesn't however solve what i believe to be the underlying issue, see no. 2

2. If using Gatsby with default SSG, you can only trigger a content refresh calling the conditionally exposed [refresh endpoint](https://www.gatsbyjs.com/docs/refreshing-content/)
    - this means that changing the event handler of StoryblokBridge will not help when you refresh the window for any reason or navigate away and back in Storyblok UI etc.; you get the original content until a refresh is registered by Gatsby dev server
    
I hope this adds at least a little bit more context to the issue 😅. Not sure what a good resolution would look like, but as @yvonnetangsu said, you would expect the integration to work live preview and save events.

**Comment by schabibi1** [comment]
Hi @yvonnetangsu and @petr166 thanks for the feedback.

I assume you two got the point I mentioned in the previous comment:

> Please keep in mind that if you serve content with SSG, the most updated content will be loaded at build time. If you'd like to load the most updated content changes whenever you hit the Save button on the Storyblok visual editor, serving content with SSR will allow you to do so.

And based on my point, would still like to have the resilient real-time live editing feature with SSG, right?
I agree and believe this FR would improve user experience a lot.

Yet, I'd like to mention the example from Next.js Preview Mode.
Next.js provides a built-in preview feature called [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode). Users can see the most updated content with SSG if they [configure](https://www.storyblok.com/faq/next-js-preview-iframes) this feature without triggering re-build.

Unfortunately, Gatsby doesn't have a similar feature except enabling [ENABLE_GATSBY_REFRESH_ENDPOINT environment variable](https://www.gatsbyjs.com/docs/how-to/local-development/environment-variables/#reserved-environment-variables) in the providers. 
> See Gatsby docs here: [https://www.gatsbyjs.com/docs/how-to/local-development/running-a-gatsby-preview-server/](https://www.gatsbyjs.com/docs/how-to/local-development/running-a-gatsby-preview-server/)

From what I see how Next.js offers, the FR for this case more likely belongs to the framework side, in my opinion.

I'll still do my best to find any possibility as I'm aware there have been a few requests about that.

**Comment by yvonnetangsu** [comment]
@schabibi1 Thank you for the follow up! Yes, if possible, we still would like to put in a FR for this to work with SSG. Our team do have an enterprise contract with Storyblok which does come with 2 hours of dev time a month, and we've never used it since we signed up 😄 If we could utilize the dev time to prioritize this FR (or even to provide a workaround), that'd be awesome. Please let me know if we need to provide you with more info. Thanks!

**Comment by schabibi1** [comment]
Hi @yvonnetangsu Thanks for the response. Really appreciate your interest in this topic.

I had discussions several times with my colleagues who also work in Storyblok JavaScript SDKs with me.
As I wrote the response [here](https://discord.com/channels/700316478792138842/879379916519510177/1083080280287498370) at Discord, we came to the conclusion that the solution belongs to the framework side as that's how other frameworks are offering. ([Next.js: Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode), [Nuxt: SSR/SPA mode](https://nuxtjs.org/docs/configuration-glossary/configuration-mode/))

Have you considered taking the official Gatsby solution by running a Gatsby dev server on a server with [Gatsby Preview Server](https://www.gatsbyjs.com/docs/how-to/local-development/running-a-gatsby-preview-server/)?

The other reason why we came up with this answer is that even if we force to make the bridge work from SDK in an SSG site, still the first page load will only have the updated version of the content when users generate the site again.

The good side of the related news is that we have planned tutorials for Gatsby to cover that with Gatsby Preview Server. What I can do is prioritize this task for this issue.
I hope it made sense.

**Comment by yvonnetangsu** [comment]
@schabibi1 Thank you so much for looking into this. I'll look into the Gatsby preview server and talk to my team about this. It would be awesome if you could please prioritize this tutorial 😄 💯 

I also saw a user (tim-refactored) posted on Discord and wonder if this is possible as a workaround:
```
I had a problem with my page loading the old content until I started typing as well. My current workaround was to skip the gatsby-source-storyblok plugin and implement a new bridge implementation similiar to how the v1 bridge works with the @storyblok/js library. I could then tap into the "enterEditMode" event of the bridge to load the initial page data.
I don't know that this is the best solution but it works for me. I have not used the ENABLE_GATSBY_REFRESH_ENDPOINT yet so I don't have any other ideas yet.
```
Tim also provided a link to the refactored hook:
https://github.com/tim-c-refactored/refactored-storyblok-hook

I haven't looked deep into how the bridge v1 works different compared to v2 with Gatsby. Is there a reason why v2 doesn't work the same as v1 (previously using v1, when we hit Save in dev mode, the visual editor preserved the latest content) - is it due to changes on Gatsby's side?

**Comment by yvonnetangsu** [comment]
@petr166 Thank you for posting your temporary solution 😄 ! We'll give that a try.

**Comment by tl-NASEEBULLAH-AHMADI** [comment]
This is a huge red flag for using Gatsby IMO. I'm facing the same issue here and currently, even the hack doesn't work unless I manually do a browser refresh.

**Comment by MaltheBlichfeldt** [comment]
Any news on this, I'm currently making a site for a client and this issue honestly makes the StoryBlok / Gatsby combination unusable...  Every time you click "save" the visual editor reverts the changes back to the "last published" changes, giving everyone that do not have this technical knowledge (eg. clients) the impression that their changes have been deleted...

**Comment by yvonnetangsu** [comment]
@schabibi1 any updates on this issue? Is it possible for us to get a timeframe? :) This completely takes away the option for non developers (content editors for example) to update their Storyblok site in the visual editor 🤔

**Comment by Patrickodey360coder** [comment]
@MaltheBlichfeldt For you to see changes, Gatsby requires you to restart your development server. To avoid repeating this step, you can use "[ENABLE_GATSBY_REFRESH_ENDPOINT environment variable]" https://www.gatsbyjs.com/docs/how-to/local-development/environment-variables/#reserved-environment-variables.

**Comment by schabibi1** [comment]
We released this FR in [gatsby-source-storyblok 6.1.1](https://github.com/storyblok/gatsby-source-storyblok/releases/tag/v6.1.1)

Closing this issue as we implemented and released it as a new feature.
